src/VecSim/spaces/computer/computer.h

```diff
@@ -0,0 +1,357 @@
+    }
+
+private:
+    MemoryUtils::unique_blob storage_blob;
+    MemoryUtils::unique_blob query_blob;
+};
+
+template <typename DistType>
+class IndexComputerAbstract : public VecsimBaseObject {
+public:
+    IndexComputerAbstract(std::shared_ptr<VecSimAllocator> allocator)
+        : VecsimBaseObject(allocator) {}
+
+    virtual ProcessedBlobs preprocess(const void *blob, size_t processed_bytes_count) const = 0;
+
+    virtual MemoryUtils::unique_blob preprocessForStorage(const void *blob,
+                                                          size_t processed_bytes_count) const = 0;
+
+    virtual MemoryUtils::unique_blob preprocessQuery(const void *blob,
+                                                     size_t processed_bytes_count) const = 0;
+
+    virtual DistType calcDistance(const void *v1, const void *v2, size_t dim) const = 0;
+    // TODO: remove alignment once datablock is implemented in HNSW
+    virtual unsigned char getAlignment() const = 0;
+};
+
+template <typename DistType, typename DistFuncType>
+class IndexComputerBasic : public IndexComputerAbstract<DistType> {
+public:
+    IndexComputerBasic(
+        std::shared_ptr<VecSimAllocator> allocator, unsigned char alignment,
+        DistanceCalculatorAbstract<DistType, DistFuncType> *distance_calculator = nullptr)
+        : IndexComputerAbstract<DistType>(allocator), alignment(alignment),
+          distance_calculator(distance_calculator) {}
+
+    ~IndexComputerBasic() override {
+        if (distance_calculator) {
+            delete distance_calculator;
+        }
+    }
+
+    virtual ProcessedBlobs preprocess(const void *original_blob,
+                                      size_t processed_bytes_count) const override;
+
+    virtual MemoryUtils::unique_blob
+    preprocessForStorage(const void *original_blob, size_t processed_bytes_count) const override;
+
+    virtual unsigned char getAlignment() const override { return alignment; }
+
+    virtual MemoryUtils::unique_blob preprocessQuery(const void *original_blob,
+                                                     size_t processed_bytes_count) const override;
+
+    DistType calcDistance(const void *v1, const void *v2, size_t dim) const override {
+        assert(this->distance_calculator);
+        return this->distance_calculator->calcDistance(v1, v2, dim);
+    }
+
+protected:
+    const unsigned char alignment;
+    DistanceCalculatorAbstract<DistType, DistFuncType> *distance_calculator;
+
+    // Allocate and copy the blob only if the original blob is not aligned.
+    MemoryUtils::unique_blob maybeCopyToAlignedMem(const void *original_blob,
+                                                   size_t blob_bytes_count) const;
+};
+
+template <typename DistType, typename DistFuncType>
+class IndexComputerExtended : public IndexComputerBasic<DistType, DistFuncType> {
+protected:
+    PreprocessorAbstract **preprocessors;
+    const size_t n_preprocessors;
```

Comment:
Vector as an overkill since we know at compile time the size of the allocation
The only alternative is to use `n_preprocessors` as a template argument